distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,782 @@
+            self._default_start += n_read
+            self._frame_nbytes_needed -= n_read
+
+    def _reset_default_buffer(self):
+        start = self._default_start
+        end = self._default_end
+
+        if start < end and start != 0:
+            self._default_buffer[: end - start] = self._default_buffer[start:end]
+            self._default_start = 0
+            self._default_end = end - start
+        elif start == end:
+            self._default_start = 0
+            self._default_end = 0
+
+    def _message_completed(self):
+        self._queue.put_nowait(self._frames)
+        self._nframes = None
+        self._frames = None
+        self._frame_lengths = None
+        self._frame_nbytes_remaining = 0
+
+    def connection_lost(self, exc=None):
+        if exc is None:
+            exc = CommClosedError("Connection closed")
+        self._exception = exc
+        self._transport = None
+        self._is_closed.set_result(None)
+
+        # Unblock read, if any
+        self._queue.put_nowait(_COMM_CLOSED)
+
+        # Unblock write, if any
+        if self._paused:
+            waiter = self._drain_waiter
+            if waiter is not None:
+                self._drain_waiter = None
+                if not waiter.done():
+                    waiter.set_exception(exc)
+
+    def pause_writing(self):
+        self._paused = True
+
+    def resume_writing(self):
+        self._paused = False
+
+        waiter = self._drain_waiter
+        if waiter is not None:
+            self._drain_waiter = None
+            if not waiter.done():
+                waiter.set_result(None)
+
+    async def read(self):
+        # Even if comm is closed, we still yield all received data before
+        # erroring
+        if self._queue is not None:
+            out = await self._queue.get()
+            if out is not _COMM_CLOSED:
+                return out
+            self._queue = None
+        raise self._exception
+
+    async def write(self, frames):
+        if self._exception:
+            raise self._exception
+
+        nframes = len(frames)
+        frames_nbytes = [nbytes(f) for f in frames]
+        # TODO: the old TCP comm included an extra `msg_nbytes` prefix that
+        # isn't really needed. We include it here for backwards compatibility,
+        # but this could be removed if we ever want to make another breaking
+        # change to the comms.
+        msg_nbytes = sum(frames_nbytes) + (nframes + 1) * 8
+        header = struct.pack(f"{nframes + 2}Q", msg_nbytes, nframes, *frames_nbytes)
+        frames = [header, *frames]
+
+        if msg_nbytes < 2 ** 17:  # 128kiB
+            # small enough, send in one go
+            frames = [b"".join(frames)]
+
+        if len(frames) > 1:
+            self._transport.writelines(frames)
+        else:
+            self._transport.write(frames[0])
+        if self._transport.is_closing():
+            await asyncio.sleep(0)
```

Comment:
Why?

Reply:
My guess is this is to make sure everything is flushed, but yeah would be curious to know more. A comment here would be good in any case

Reply:
This was a holdover from work on `ery`, but doesn't actually apply to our comm usage patterns (which are mostly request-response based). If you're doing multiple sends without a single read, there's a chance that no `await` call will ever yield to the event loop. This can prevent event loop callbacks like `connection_lost` from ever being called, meaning the writer never notices that the connection is in the process of closing. In this case we need to manually yield control so the callback can run properly.

Since our usage patterns will never hit this (the process would have to write infinitely without ever yielding), I've just removed this call.